### PR TITLE
GOV 223 Update batch API

### PR DESCRIPTION
**Pull request by danishjamal104**
## Description

* [Added migration script for adding fields in batches table.](https://github.com/openMF/ph-ee-operations-app/commit/d6e6253c2e114a1190c60fe40b3ee40a4a00ffd0)
* [Updated the DTO to support new DB structure for batch table.](https://github.com/openMF/ph-ee-operations-app/pull/57/commits/9305235de8eda571ba813f7f185837ea016b8306)
* [Added new api /batch with offset, limit, sort and date filter.](https://github.com/openMF/ph-ee-operations-app/pull/57/commits/2f08fdf3ed2aa95b0f6cff304672b0ec749a7dbe)

[GOV-223](https://mifosforge.jira.com/browse/GOV-223)

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [X] Design related bullet points or design document link related to this PR added in the description above. 

- [X] Updated corresponding Postman Collection or Api documentation for the changes in this PR.

- [X] Create/update unit or integration tests for verifying the changes made.

- [X] Add required Swagger annotation and update API documentation with details of any API changes if applicable

- [X] Followed the naming conventions as given in https://docs.google.com/document/d/1Q4vaMSzrTxxh9TS0RILuNkSkYCxotuYk1Xe0CMIkkCU/edit?usp=sharing

[GOV-223]: https://mifosforge.jira.com/browse/GOV-223?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/org/apache/fineract/api/BatchApi.java [hunk 1]
```diff
@@ -63,6 +76,76 @@ public Page<Batch> getBatches(@RequestParam(value = "page", required = false, de
         }
 
         return batchRepository.findAll(specifications, pager);
+    }*/
```
Comment (arkadasfynarfin) [inline src/main/java/org/apache/fineract/api/BatchApi.java:79]: @danishjamal104 dead code?

File: src/main/java/org/apache/fineract/api/BatchApi.java [hunk 2]
```diff
@@ -44,10 +51,16 @@ public class BatchApi {
...
+    private DateUtil dateUtil;
+
+    @Autowired
+    private ObjectMapper objectMapper;
+
     @Value("${application.bucket-name}")
     private String bucketName;
 
-    @GetMapping("/batches")
+    /*@GetMapping("/batches")
```
Comment (arkadasfynarfin) [inline src/main/java/org/apache/fineract/api/BatchApi.java:63]: @danishjamal104 dead code. If not needed, can we remove this?

File: src/main/java/org/apache/fineract/api/BatchApi.java [hunk 3]
```diff
@@ -63,6 +76,76 @@ public Page<Batch> getBatches(@RequestParam(value = "page", required = false, de
         }
 
         return batchRepository.findAll(specifications, pager);
+    }*/
+
+    @GetMapping("/batches")
+    public String getBatch123(@RequestParam(value = "offset", required = false, defaultValue = "0") Integer offset,
```
Comment (arkadasfynarfin) [inline src/main/java/org/apache/fineract/api/BatchApi.java:82]: @danishjamal104 Use descriptive names for method. Even `getBatch` is vague.

File: src/main/java/org/apache/fineract/api/BatchApi.java [hunk 4]
```diff
@@ -63,6 +76,76 @@ public Page<Batch> getBatches(@RequestParam(value = "page", required = false, de
...
+                                   @RequestParam(value = "limit", required = false, defaultValue = "10") Integer limit,
+                                   @RequestParam(value = "sort", required = false, defaultValue = "+completedAt")
+                                       String sort,
+                                @RequestParam(value = "dateFrom", required = false) String startFrom,
+                                @RequestParam(value = "dateTo", required = false) String startTo) {
+
+        List<Specification<Batch>> specifications = new ArrayList<>();
```
Comment (arkadasfynarfin) [inline src/main/java/org/apache/fineract/api/BatchApi.java:89]: @danishjamal104 Method too long. Break it into 2 or more methods.

File: src/main/java/org/apache/fineract/api/BatchApi.java [hunk 5]
```diff
@@ -21,9 +26,11 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Optional;
-
+import java.util.ArrayList;
+import static org.apache.fineract.core.service.OperatorUtils.dateFormat;
```
Comment (arkadasfynarfin) [inline src/main/java/org/apache/fineract/api/BatchApi.java:30]: @danishjamal104 Static imports are usually at the top. Checkstyle your code.

**Review by fynmanoj — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/fineract/api/BatchApi.java (hunk 2 above)
Comment (danishjamal104) [reply src/main/java/org/apache/fineract/api/BatchApi.java]: resolved

File: src/main/java/org/apache/fineract/api/BatchApi.java [hunk 6]
```diff
@@ -63,6 +76,76 @@ public Page<Batch> getBatches(@RequestParam(value = "page", required = false, de
...
+        try {
+            if (startFrom != null && startTo != null) {
+                specifications.add(BatchSpecs.between(Batch_.startedAt, dateFormat().parse(startFrom), dateFormat().parse(startTo)));
+            } else if (startFrom != null) {
+                specifications.add(BatchSpecs.later(Batch_.startedAt, dateFormat().parse(startFrom)));
+            } else if (startTo != null) {
+                specifications.add(BatchSpecs.earlier(Batch_.startedAt, dateFormat().parse(startTo)));
+            }
+        } catch (Exception e) {
+            log.warn("failed to parse dates {} / {}", startFrom, startTo);
```
[the remark this answers is not in the document]
Comment (danishjamal104) [reply src/main/java/org/apache/fineract/api/BatchApi.java]: resolved

File: src/main/java/org/apache/fineract/operations/BatchPaginatedResponse.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.List;
+
+@Getter
+@Setter
+public class BatchPaginatedResponse {
+
+    long totalBatches, totalTransactions, totalAmount, totalApprovedCount, totalApprovedAmount, totalSubBatchesCreated;
+    List<Batch> data;
+
+    public void setTotalBatches(Long totalBatches) {
```
Comment (fynmanoj) [inline src/main/java/org/apache/fineract/operations/BatchPaginatedResponse.java:15]: Lets use builder pattern here to construct the object in a better way

File: src/main/java/org/apache/fineract/api/BatchApi.java [hunk 8]
```diff
@@ -44,25 +59,82 @@ public class BatchApi {
...
+            sortedBy = sort;
         }
+        sortedBy = sortedBy.replace(" ", "");
+        log.info("Sorting by: {} and Sorting direction: {}", sortedBy, sortDirection.name());
+        return new Sort(sortDirection, sortedBy);
+    }
 
-        return batchRepository.findAll(specifications, pager);
+    @GetMapping("/batches")
+    public BatchPaginatedResponse getBatch(@RequestParam(value = "offset", required = false, defaultValue = "0")
```
Comment (fynmanoj) [inline src/main/java/org/apache/fineract/api/BatchApi.java:93]: Please change/ check documentation for this API ( Postman and Swagger)

**Review by fynmanoj — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/fineract/api/BatchApi.java (hunk 5 above)
Comment (danishjamal104) [reply src/main/java/org/apache/fineract/api/BatchApi.java]: FIXED

File: .circleci/config.yml [hunk 9]
```diff
@@ -2,7 +2,7 @@ version: 2.1
 executors:
   docker-executor:
     docker:
-      - image: circleci/openjdk:17-buster-node-browsers-legacy
+      - image: circleci/openjdk:8-buster-node-browsers-legacy
```
[the remark this answers is not in the document]
Comment (danishjamal104) [reply .circleci/config.yml]: done

File: src/main/resources/sql/migrations/tenant/V43__add_busget_account_field_in_batch.sql [hunk 10]
```diff
@@ -0,0 +1,8 @@
+ALTER TABLE `batches`
```
[the remark this answers is not in the document]
Comment (danishjamal104) [reply src/main/resources/sql/migrations/tenant/V43__add_busget_account_field_in_batch.sql]: fixed

File: src/main/java/org/apache/fineract/service/BatchDbServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        List<Batch> batches = batchRepository.findAllBatch(registeringInstitutionId, payerFsp, batchId, pager);
+        return getBatchPaginatedResponseInstance(totalBatches.orElse(0L), totalTransactions.orElse(0L),
+                totalAmount.orElse(0L), totalApprovedCount.orElse(0L),
+                totalApprovedAmount.orElse(0L), 10, batches);
+    }
+
+    @Override
+    public BatchPaginatedResponse getBatchDateTo(String startTo, String registeringInstitutionId, String payerFsp, String batchId, PageRequest pager) {
+        try {
+            Optional<Long> totalTransactions = batchRepository.getTotalTransactionsDateTo(
```
[the remark this answers is not in the document]
Comment (danishjamal104) [reply src/main/java/org/apache/fineract/service/BatchDbServiceImpl.java]: updated and tested

File: src/main/java/org/apache/fineract/service/BatchDbService.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+    public BatchPaginatedResponse getBatch(String startFrom, String startTo, String registeringInstitutionId,
+                                           String payerFsp, String batchId, PageRequest pager);
+
+    public BatchPaginatedResponse getBatch(String registeringInstitutionId, String payerFsp, String batchId, PageRequest pager);
+
+    public BatchPaginatedResponse getBatchDateTo(String startTo, String registeringInstitutionId,
+                                           String payerFsp, String batchId, PageRequest pager);
+
+    public BatchPaginatedResponse getBatchDateFrom(String startFrom, String registeringInstitutionId,
```
[the remark this answers is not in the document]
Comment (danishjamal104) [reply src/main/java/org/apache/fineract/service/BatchDbService.java:18]: not part of the story will do it once this PR is merged.

File: src/main/java/org/apache/fineract/service/BatchDbServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+        } catch (Exception e) {
+            log.warn("failed to parse startTo date {}", startTo);
+            return null;
+        }
+    }
+
+    @Override
+    public BatchPaginatedResponse getBatchDateFrom(String startFrom, String registeringInstitutionId, String payerFsp, String batchId, PageRequest pager) {
+        try {
+            Optional<Long> totalTransactions = batchRepository.getTotalTransactionsDateFrom(
```
[the remark this answers is not in the document]
Comment (danishjamal104) [reply src/main/java/org/apache/fineract/service/BatchDbServiceImpl.java]: done

File: src/main/java/org/apache/fineract/api/BatchApi.java [hunk 14]
```diff
@@ -63,6 +76,76 @@ public Page<Batch> getBatches(@RequestParam(value = "page", required = false, de
...
+            sortedBy = sort.split("\\+")[1];
+        } else if (sort.contains("-") && sort.split("-").length == 2) {
+			sortDirection = Sort.Direction.DESC;
+            sortedBy = sort.split("-")[1];
+        } else {
+            sortDirection = Sort.Direction.ASC;
+            sortedBy = sort;
+        }
+        sortedBy = sortedBy.replace(" ", "");
+        log.info("Sorting by: {} and Sorting direction: {}", sortedBy, sortDirection.name());
```
[the remark this answers is not in the document]
Comment (danishjamal104) [reply src/main/java/org/apache/fineract/api/BatchApi.java]: older comment already updated the code.

File: src/main/java/org/apache/fineract/service/BatchDbService.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (danishjamal104) [reply src/main/java/org/apache/fineract/service/BatchDbService.java:18]: done

**Review by fynmanoj — APPROVED** [review]

**Review by fynmanoj — APPROVED** [review]

**Pull request merged**